### PR TITLE
Tracks: Add tracking event when OBW is started

### PR DESCRIPTION
**Pull request by joshuatf**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds a tracking event when the OBW is started.

*Question*: Will this event ever really get tracked?  `woocommerce_allow_tracking` is not set until after the first step, so this code probably won't get run unless there is another place this can be activated first.

### How to test the changes in this Pull Request:

1. Visit the initial OBW page `/wp-admin/admin.php?page=wc-setup`.
2. Check that the event is fired.
3. Visit the next step and and click "Store setup" to return to the first step.
4. Make sure the event is not tracked again.

### CODE REVIEW

**Comment by claudiosanches** [comment]
> _Question_: Will this event ever really get tracked? `woocommerce_allow_tracking` is not set until after the first step, so this code probably won't get run unless there is another place this can be activated first.

Yep, sounds like something could be do in two ways, to also start it when update this option in that page.

**Review by psealock — APPROVED** [review]
LGTM

**Comment by psealock** [comment]
Whoops I approved by accident, although the code looks good.

The tracking is initialized by an explicit check for `yes`, but maybe we can hook this event in for an undefined value?

**Comment by joshuatf** [comment]
@psealock @claudiosanches Changed this to check when the option is added for tracking.  Had to move this method into `WC_Site_Tracking` to bypass the `is_tracking_enabled()` check.

Let me know if that looks a bit better.

**To Test**:
1.  Delete the option `woocommerce_allow_tracking` from the options table.
2.  Go to `wp-admin/admin.php?page=wc-setup`.
3.  Opt in to tracking and continue to the next step.
4.  Check that the event is tracked.
5.  Repeat this process, but opt out of tracking and make sure the event is not tracked.

File: includes/tracks/class-wc-site-tracking.php [hunk 1]
```diff
@@ -21,8 +21,20 @@ public static function is_tracking_enabled() {
...
+		if (
+			! apply_filters( 'woocommerce_apply_user_tracking', true ) ||
+			(
+				'yes' !== get_option( 'woocommerce_allow_tracking' ) &&
+				// Check if tracking is actively being opted into.
+				// phpcs:disable WordPress.Security.NonceVerification.NoNonceVerification, WordPress.Security.ValidatedSanitizedInput
+				(
+					! isset( $_POST['wc_tracker_checkbox'] ) ||
+					'yes' !== sanitize_text_field( $_POST['wc_tracker_checkbox'] )
+				)
```
Comment (psealock) [inline includes/tracks/class-wc-site-tracking.php]: Nice solution here! Its a bit hard to parse. How about something like this?

```php
if ( ! apply_filters( 'woocommerce_apply_user_tracking', true ) ) {
    return false;
}

// Check if tracking is actively being opted into.
$is_obw_opting_in = isset( $_POST['wc_tracker_checkbox'] ) && 'yes' === sanitize_text_field( $_POST['wc_tracker_checkbox'] );

if ( 'yes' !== get_option( 'woocommerce_allow_tracking' ) && ! $is_obw_opting_in ) {
   return false;
}
```

File: includes/tracks/events/class-wc-admin-setup-wizard-tracking.php [hunk 2]
```diff
@@ -151,7 +167,7 @@ public static function track_recommended() {
 	 *
 	 * @return void
 	 */
-	public static function track_activate() {
+	public static function track_jetpack_activate() {
```
Comment (psealock) [inline includes/tracks/events/class-wc-admin-setup-wizard-tracking.php]: 👍 Thanks!

File: includes/tracks/events/class-wc-admin-setup-wizard-tracking.php [hunk 3]
```diff
@@ -19,9 +19,25 @@ public static function init() {
...
 	}
 
+	/**
+	 * Track when tracking is opted into and OBW has started.
+	 *
+	 * @param string $option Option name.
+	 * @param string $value  Option value.
+	 * @return void
+	 */
+	public static function track_start( $option, $value ) {
```
Comment (psealock) [inline includes/tracks/events/class-wc-admin-setup-wizard-tracking.php]: I can't find documentation on this action. Will `$option` always be the correct option name? I assume so, but figured I'd ask.

**Review by psealock — APPROVED** [review]
Testing great and a nice solution to the question of tracking while actively opting in. Just a couple comments on the code, but looking good.

File: includes/tracks/class-wc-site-tracking.php (hunk 1 above)
Comment (joshuatf) [reply includes/tracks/class-wc-site-tracking.php]: Seems like a good way to handle this and much more readable - thanks! 
 Added in https://github.com/woocommerce/woocommerce/pull/22877/commits/848504222d76a65c09e8946ce30c3051a11c3cb0

File: includes/tracks/events/class-wc-admin-setup-wizard-tracking.php (hunk 3 above)
Comment (joshuatf) [reply includes/tracks/events/class-wc-admin-setup-wizard-tracking.php]: Yep, `$option` will always be correct.  Seems a bit redundant, but there's probably a use case for a generic function that handles a bunch of options.
